File: crates/cnc-render-contracts/src/units.rs
use serde::{Deserialize, Serialize};

use crate::validation::{ContractResult, require_finite, require_positive};

pub const MILLIMETERS_PER_INCH: f64 = 25.4;
pub const SECONDS_PER_MINUTE: f64 = 60.0;
pub const RADIANS_PER_DEGREE: f64 = std::f64::consts::PI / 180.0;

macro_rules! finite_quantity {
    ($name:ident) => {
        #[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
        #[serde(transparent)]
        pub struct $name(f64);

        impl $name {
            pub fn try_new(value: f64) -> ContractResult<Self> {
                require_finite(value, stringify!($name)).map(Self)
            }

            pub const fn get(self) -> f64 {
                self.0
            }
        }
    };
}

macro_rules! positive_quantity {
    ($name:ident) => {
        #[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
        #[serde(transparent)]
        pub struct $name(f64);

        impl $name {
            pub fn try_new(value: f64) -> ContractResult<Self> {
                require_positive(value, stringify!($name)).map(Self)
            }

            pub const fn get(self) -> f64 {
                self.0
            }
        }
    };
}

finite_quantity!(CoordinateMm);
finite_quantity!(AngleRad);
finite_quantity!(TimeS);
positive_quantity!(LengthMm);
positive_quantity!(SpindleSpeedRpm);
positive_quantity!(FeedMmPerMin);
positive_quantity!(FeedMmPerRev);
positive_quantity!(FeedMmPerTooth);

pub fn millimeters_to_inches(millimeters: f64) -> ContractResult<f64> {
    Ok(require_finite(millimeters, "millimeters")? / MILLIMETERS_PER_INCH)
}

pub fn inches_to_millimeters(inches: f64) -> ContractResult<f64> {
    Ok(require_finite(inches, "inches")? * MILLIMETERS_PER_INCH)
}

pub fn millimeters_per_minute_to_inches_per_minute(
    millimeters_per_minute: f64,
) -> ContractResult<f64> {
    Ok(require_finite(millimeters_per_minute, "millimetersPerMinute")? / MILLIMETERS_PER_INCH)
}

pub fn inches_per_minute_to_millimeters_per_minute(inches_per_minute: f64) -> ContractResult<f64> {
    Ok(require_finite(inches_per_minute, "inchesPerMinute")? * MILLIMETERS_PER_INCH)
}

pub fn millimeters_per_revolution_to_inches_per_revolution(
    millimeters_per_revolution: f64,
) -> ContractResult<f64> {
    Ok(
        require_finite(millimeters_per_revolution, "millimetersPerRevolution")?
            / MILLIMETERS_PER_INCH,
    )
}

pub fn inches_per_revolution_to_millimeters_per_revolution(
    inches_per_revolution: f64,
) -> ContractResult<f64> {
    Ok(require_finite(inches_per_revolution, "inchesPerRevolution")? * MILLIMETERS_PER_INCH)
}

pub fn millimeters_per_tooth_to_inches_per_tooth(
    millimeters_per_tooth: f64,
) -> ContractResult<f64> {
    Ok(require_finite(millimeters_per_tooth, "millimetersPerTooth")? / MILLIMETERS_PER_INCH)
}

pub fn inches_per_tooth_to_millimeters_per_tooth(inches_per_tooth: f64) -> ContractResult<f64> {
    Ok(require_finite(inches_per_tooth, "inchesPerTooth")? * MILLIMETERS_PER_INCH)
}

pub fn revolutions_per_minute_to_revolutions_per_second(
    revolutions_per_minute: f64,
) -> ContractResult<f64> {
    Ok(require_finite(revolutions_per_minute, "revolutionsPerMinute")? / SECONDS_PER_MINUTE)
}

pub fn revolutions_per_second_to_revolutions_per_minute(
    revolutions_per_second: f64,
) -> ContractResult<f64> {
    Ok(require_finite(revolutions_per_second, "revolutionsPerSecond")? * SECONDS_PER_MINUTE)
}

pub fn degrees_to_radians(degrees: f64) -> ContractResult<f64> {
    Ok(require_finite(degrees, "degrees")? * RADIANS_PER_DEGREE)
}

pub fn radians_to_degrees(radians: f64) -> ContractResult<f64> {
    Ok(require_finite(radians, "radians")? / RADIANS_PER_DEGREE)
}

pub fn round_for_display(value: f64, decimal_places: u32) -> ContractResult<f64> {
    let value = require_finite(value, "displayValue")?;
    let exponent = i32::try_from(decimal_places).unwrap_or(i32::MAX);
    let factor = 10_f64.powi(exponent);

    if !factor.is_finite() {
        return Ok(value);
    }

    Ok((value * factor).round() / factor)
}

#[cfg(test)]
mod tests {
    use serde::Deserialize;

    use super::{
        CoordinateMm, FeedMmPerMin, LengthMm, SpindleSpeedRpm, degrees_to_radians,
        inches_per_minute_to_millimeters_per_minute,
        inches_per_revolution_to_millimeters_per_revolution,
        inches_per_tooth_to_millimeters_per_tooth, inches_to_millimeters,
        millimeters_per_minute_to_inches_per_minute,
        millimeters_per_revolution_to_inches_per_revolution,
        millimeters_per_tooth_to_inches_per_tooth, millimeters_to_inches, radians_to_degrees,
        revolutions_per_minute_to_revolutions_per_second,
        revolutions_per_second_to_revolutions_per_minute, round_for_display,
    };

    #[derive(Debug, Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct GoldenUnits {
        feeds: GoldenFeeds,
        absolute_tolerance: f64,
        relative_tolerance: f64,
    }

    #[derive(Debug, Deserialize)]
    #[serde(rename_all = "camelCase")]
    struct GoldenFeeds {
        rpm: f64,
        mm_per_min: f64,
        in_per_min: f64,
        mm_per_rev: f64,
        in_per_rev: f64,
        mm_per_tooth: f64,
        in_per_tooth: f64,
    }

    fn golden_units() -> GoldenUnits {
        serde_json::from_str(include_str!("../../../tests/fixtures/m1/units.golden.json"))
            .expect("units golden fixture")
    }

    fn approximately_equal(
        left: f64,
        right: f64,
        absolute_tolerance: f64,
        relative_tolerance: f64,
    ) -> bool {
        let absolute = (left - right).abs();
        let relative = absolute / left.abs().max(right.abs()).max(1.0);
        absolute <= absolute_tolerance || relative <= relative_tolerance
    }

    #[test]
    fn units_length_golden_round_trip() {
        let inches = millimeters_to_inches(25.4).expect("finite millimeters");
        assert!(approximately_equal(inches, 1.0, 1.0e-12, 1.0e-12));
        let millimeters = inches_to_millimeters(inches).expect("finite inches");
        assert!(approximately_equal(millimeters, 25.4, 1.0e-12, 1.0e-12));
    }

    #[test]
    fn units_angle_golden_round_trip() {
        let radians = degrees_to_radians(180.0).expect("finite degrees");
        assert!(approximately_equal(
            radians,
            std::f64::consts::PI,
            1.0e-12,
            1.0e-12
        ));
        let degrees = radians_to_degrees(radians).expect("finite radians");
        assert!(approximately_equal(degrees, 180.0, 1.0e-12, 1.0e-12));
    }

    #[test]
    fn units_feed_and_spindle_golden_round_trips() {
        let golden = golden_units();
        let feeds = golden.feeds;
        let absolute = golden.absolute_tolerance;
        let relative = golden.relative_tolerance;

        let in_per_min = millimeters_per_minute_to_inches_per_minute(feeds.mm_per_min)
            .expect("mm/min conversion");
        assert!(approximately_equal(
            in_per_min,
            feeds.in_per_min,
            absolute,
            relative
        ));
        assert!(approximately_equal(
            inches_per_minute_to_millimeters_per_minute(in_per_min).expect("in/min conversion"),
            feeds.mm_per_min,
            absolute,
            relative
        ));

        let in_per_rev = millimeters_per_revolution_to_inches_per_revolution(feeds.mm_per_rev)
            .expect("mm/rev conversion");
        assert!(approximately_equal(
            in_per_rev,
            feeds.in_per_rev,
            absolute,
            relative
        ));
        assert!(approximately_equal(
            inches_per_revolution_to_millimeters_per_revolution(in_per_rev)
                .expect("in/rev conversion"),
            feeds.mm_per_rev,
            absolute,
            relative
        ));

        let in_per_tooth = millimeters_per_tooth_to_inches_per_tooth(feeds.mm_per_tooth)
            .expect("mm/tooth conversion");
        assert!(approximately_equal(
            in_per_tooth,
            feeds.in_per_tooth,
            absolute,
            relative
        ));
        assert!(approximately_equal(
            inches_per_tooth_to_millimeters_per_tooth(in_per_tooth).expect("in/tooth conversion"),
            feeds.mm_per_tooth,
            absolute,
            relative
        ));

        let revolutions_per_second =
            revolutions_per_minute_to_revolutions_per_second(feeds.rpm).expect("rpm conversion");
        assert!(approximately_equal(
            revolutions_per_second,
            100.0,
            absolute,
            relative
        ));
        assert!(approximately_equal(
            revolutions_per_second_to_revolutions_per_minute(revolutions_per_second)
                .expect("rev/s conversion"),
            feeds.rpm,
            absolute,
            relative
        ));
    }

    #[test]
    fn units_positive_quantities_reject_invalid_values() {
        assert!(LengthMm::try_new(-1.0).is_err());
        assert!(FeedMmPerMin::try_new(0.0).is_err());
        assert!(SpindleSpeedRpm::try_new(f64::INFINITY).is_err());
    }

    #[test]
    fn units_negative_zero_is_rejected() {
        let error = CoordinateMm::try_new(-0.0).expect_err("negative zero must fail");
        assert_eq!(error.code, "number.negative_zero");
        assert!(millimeters_to_inches(-0.0).is_err());
        assert!(round_for_display(-0.0, 3).is_err());
    }

    #[test]
    fn units_display_rounding_does_not_mutate_internal_precision() {
        let internal = 12.345_678_901_234_5;
        let displayed = round_for_display(internal, 3).expect("finite value");
        assert_eq!(displayed, 12.346);
        assert_eq!(internal, 12.345_678_901_234_5);
    }
}

File: packages/contracts/src/units.ts
const MILLIMETERS_PER_INCH = 25.4;
const SECONDS_PER_MINUTE = 60;
const RADIANS_PER_DEGREE = Math.PI / 180;

declare const quantityBrand: unique symbol;

export type Quantity<Unit extends string> = number & {
  readonly [quantityBrand]: Unit;
};

export type Millimeters = Quantity<"mm">;
export type Inches = Quantity<"in">;
export type Radians = Quantity<"rad">;
export type Degrees = Quantity<"deg">;
export type RevolutionsPerMinute = Quantity<"rpm">;
export type RevolutionsPerSecond = Quantity<"rev/s">;
export type MillimetersPerMinute = Quantity<"mm/min">;
export type MillimetersPerRevolution = Quantity<"mm/rev">;
export type MillimetersPerTooth = Quantity<"mm/tooth">;

export function assertFiniteNumber(value: number, label: string): number {
  if (!Number.isFinite(value) || Object.is(value, -0)) {
    throw new RangeError(`${label} must be finite and must not be negative zero`);
  }
  return value;
}

export function assertPositiveNumber(value: number, label: string): number {
  assertFiniteNumber(value, label);
  if (value <= 0) {
    throw new RangeError(`${label} must be greater than zero`);
  }
  return value;
}

export function millimeters(value: number): Millimeters {
  return assertFiniteNumber(value, "millimeters") as Millimeters;
}

export function inches(value: number): Inches {
  return assertFiniteNumber(value, "inches") as Inches;
}

export function radians(value: number): Radians {
  return assertFiniteNumber(value, "radians") as Radians;
}

export function degrees(value: number): Degrees {
  return assertFiniteNumber(value, "degrees") as Degrees;
}

export function revolutionsPerMinute(value: number): RevolutionsPerMinute {
  return assertPositiveNumber(value, "spindle speed rpm") as RevolutionsPerMinute;
}

export function revolutionsPerSecond(value: number): RevolutionsPerSecond {
  return assertPositiveNumber(value, "spindle speed rev/s") as RevolutionsPerSecond;
}

export function millimetersPerMinute(value: number): MillimetersPerMinute {
  return assertPositiveNumber(value, "feed mm/min") as MillimetersPerMinute;
}

export function millimetersPerRevolution(
  value: number,
): MillimetersPerRevolution {
  return assertPositiveNumber(value, "feed mm/rev") as MillimetersPerRevolution;
}

export function millimetersPerTooth(value: number): MillimetersPerTooth {
  return assertPositiveNumber(value, "feed mm/tooth") as MillimetersPerTooth;
}

export function millimetersToInches(value: Millimeters): Inches {
  return inches(value / MILLIMETERS_PER_INCH);
}

export function inchesToMillimeters(value: Inches): Millimeters {
  return millimeters(value * MILLIMETERS_PER_INCH);
}

export function degreesToRadians(value: Degrees): Radians {
  return radians(value * RADIANS_PER_DEGREE);
}

export function radiansToDegrees(value: Radians): Degrees {
  return degrees(value / RADIANS_PER_DEGREE);
}

export function revolutionsPerMinuteToRevolutionsPerSecond(
  value: RevolutionsPerMinute,
): RevolutionsPerSecond {
  return revolutionsPerSecond(value / SECONDS_PER_MINUTE);
}

export function revolutionsPerSecondToRevolutionsPerMinute(
  value: RevolutionsPerSecond,
): RevolutionsPerMinute {
  return revolutionsPerMinute(value * SECONDS_PER_MINUTE);
}

export function millimetersPerMinuteToInchesPerMinute(
  value: MillimetersPerMinute,
): number {
  return assertPositiveNumber(value / MILLIMETERS_PER_INCH, "feed in/min");
}

export function inchesPerMinuteToMillimetersPerMinute(
  value: number,
): MillimetersPerMinute {
  return millimetersPerMinute(
    assertPositiveNumber(value, "feed in/min") * MILLIMETERS_PER_INCH,
  );
}

export function millimetersPerRevolutionToInchesPerRevolution(
  value: MillimetersPerRevolution,
): number {
  return assertPositiveNumber(value / MILLIMETERS_PER_INCH, "feed in/rev");
}

export function inchesPerRevolutionToMillimetersPerRevolution(
  value: number,
): MillimetersPerRevolution {
  return millimetersPerRevolution(
    assertPositiveNumber(value, "feed in/rev") * MILLIMETERS_PER_INCH,
  );
}

export function millimetersPerToothToInchesPerTooth(
  value: MillimetersPerTooth,
): number {
  return assertPositiveNumber(value / MILLIMETERS_PER_INCH, "feed in/tooth");
}

export function inchesPerToothToMillimetersPerTooth(
  value: number,
): MillimetersPerTooth {
  return millimetersPerTooth(
    assertPositiveNumber(value, "feed in/tooth") * MILLIMETERS_PER_INCH,
  );
}

export function roundForDisplay(
  canonicalValue: number,
  decimalPlaces: number,
): string {
  assertFiniteNumber(canonicalValue, "canonical display value");
  if (
    !Number.isSafeInteger(decimalPlaces) ||
    decimalPlaces < 0 ||
    decimalPlaces > 12
  ) {
    throw new RangeError("decimalPlaces must be an integer from 0 through 12");
  }
  return canonicalValue.toFixed(decimalPlaces);
}

export function quantitiesApproximatelyEqual(
  left: number,
  right: number,
  absoluteTolerance = 1e-12,
  relativeTolerance = 1e-12,
): boolean {
  assertFiniteNumber(left, "left quantity");
  assertFiniteNumber(right, "right quantity");
  const absoluteError = Math.abs(left - right);
  const relativeError =
    absoluteError / Math.max(Math.abs(left), Math.abs(right), 1);
  return (
    absoluteError <= absoluteTolerance || relativeError <= relativeTolerance
  );
}
